### ISSUE TITLE
Ubuntu 20.04 Memory Leak 

### ISSUE THREAD

**Issue by Sir-Alex-L** [datto/dattobd#294]
It seems that dattobd is not releasing RAM upon backup completion, The RAM usage jumps up with every backup taken.
The VM is running on Hyper-V.
Only fix is to reboot the VM. Tried restarting the kernel module with no difference.

Journalctl entry
```
Jan 13 10:20:08 dlad[1016]: Beginning a dla backup: 8f2e65de-4f5f-4978-a5a8-5927c92b74e8
Jan 13 10:20:15 sudo[51371]: pam_unix(sudo:session): session closed for user root
Jan 13 10:20:18 multipathd[597]: datto0: HDIO_GETGEO failed with 25
Jan 13 10:20:18 multipathd[597]: datto0: failed to get udev uid: Invalid argument
Jan 13 10:20:18 multipathd[597]: datto0: failed to get unknown uid: Invalid argument
Jan 13 10:20:21 multipathd[597]: datto1: HDIO_GETGEO failed with 25
Jan 13 10:20:21 multipathd[597]: datto1: failed to get udev uid: Invalid argument
Jan 13 10:20:21 multipathd[597]: datto1: failed to get unknown uid: Invalid argument
Jan 13 10:20:21 kernel: EXT4-fs (datto1): write access unavailable, skipping orphan cleanup
Jan 13 10:20:21 kernel: EXT4-fs (datto1): mounted filesystem without journal. Opts: norecovery
Jan 13 10:20:26 sudo[51737]:  itadmin : TTY=pts/1 ; PWD=/home/itadmin ; USER=root ; COMMAND=/usr/sbin/swapon -a
Jan 13 10:20:26 sudo[51737]: pam_unix(sudo:session): session opened for user root by itadmin(uid=0)
Jan 13 10:20:26 sudo[51737]: pam_unix(sudo:session): session closed for user root
Jan 13 10:20:26 kernel: Adding 2097148k swap on /swap.img.  Priority:-2 extents:4 across:2244604k FS
Jan 13 10:20:28 sudo[51751]:  itadmin : TTY=pts/1 ; PWD=/home/itadmin ; USER=root ; COMMAND=/usr/bin/htop
Jan 13 10:20:28 sudo[51751]: pam_unix(sudo:session): session opened for user root by itadmin(uid=0)
Jan 13 10:20:30 systemd[1]: tmp-dattoMountRoot-6919c795\x2d02ff\x2d4dcd\x2db6e8\x2d4c87d7709ba0.mount: Succeeded.
Jan 13 10:20:30 systemd[5759]: tmp-dattoMountRoot-6919c795\x2d02ff\x2d4dcd\x2db6e8\x2d4c87d7709ba0.mount: Succeeded.
Jan 13 10:20:30 systemd[5759]: tmp-dattoMountRoot-D682\x2d1F98.mount: Succeeded.
Jan 13 10:21:34 multipathd[597]: datto0: path already removed
Jan 13 10:20:30 systemd[1]: tmp-dattoMountRoot-D682\x2d1F98.mount: Succeeded.
Jan 13 10:21:34 multipathd[597]: datto1: path already removed
Jan 13 10:21:34 [1016]: Ending a dla backup: 8f2e65de-4f5f-4978-a5a8-5927c92b74e8. Result:  SUCCESSFUL!
```

RAM usage before backup

```
itadmin@flex:~$ free -h
              total        used        free      shared  buff/cache   available
Mem:          7.7Gi       4.5Gi       1.9Gi       3.0Mi       1.4Gi       2.9Gi
Swap:         2.0Gi          0B       2.0Gi
```

RAM usage after the backup

```
itadmin@flex:~$ free -h
              total        used        free      shared  buff/cache   available
Mem:          7.7Gi       6.4Gi       134Mi       3.0Mi       1.2Gi       1.1Gi
Swap:         2.0Gi       1.0Mi       2.0Gi
```

Kernel:
`Linux version 5.4.0-136-generic (buildd@lcy02-amd64-068) (gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1)) #153-Ubuntu SMP Thu Nov 24 15:56:58 UTC 2022`

Kernel log
```
Jan 13 10:20:21 flex kernel: [43067.243895] EXT4-fs (datto1): write access unavailable, skipping orphan cleanup
Jan 13 10:20:21 flex kernel: [43067.243931] EXT4-fs (datto1): mounted filesystem without journal. Opts: norecovery
Jan 13 10:38:12 flex kernel: [44134.709523] datto: failed to locate system call table, persistence disabled
Jan 13 10:45:37 flex kernel: [44580.047702] datto: failed to locate system call table, persistence disabled
Jan 13 10:54:53 flex kernel: [45135.956321] datto: failed to locate system call table, persistence disabled
Jan 13 11:09:16 flex kernel: [45999.051663] EXT4-fs (datto1): write access unavailable, skipping orphan cleanup
Jan 13 11:09:16 flex kernel: [45999.051699] EXT4-fs (datto1): mounted filesystem without journal. Opts: norecovery
Jan 13 11:19:16 flex kernel: [46599.064570] EXT4-fs (datto1): write access unavailable, skipping orphan cleanup
Jan 13 11:19:16 flex kernel: [46599.064609] EXT4-fs (datto1): mounted filesystem without journal. Opts: norecovery
```
[dba_2023-01-13.log](https://github.com/datto/dattobd/files/10414808/dba_2023-01-13.log)

Installed package version
`dlad/focal,now 2.7.1.1-1.1ubuntu20.04 amd64 [installed]`

The kernel module that get installed with above package is
```
Module:  dattobd
Version: 0.10.15
Kernel:  5.4.0-136-generic (x86_64)
```

**Comment by adnanshaheen** [comment]
Thank you for posting this. This is a known issue, and we are working on a fix.

**Comment by ScottMonolith** [comment]
@adnanshaheen any updates?  Has this issue been resolved and is still marked 'open'?

I am running the latest dlad agent, although the version shows something odd:

dpkg -l dlad
3.0.16.0-91.5ubuntu18.04

Odd because it shows ubuntu18.04, I am running Ubuntu 22.04.2.

**Comment by adnanshaheen** [comment]
I don't have any insight any longer. I am not part of the project anymore.

**Comment by Sir-Alex-L** [comment]
According to Datto article it's supposed to be supported. 
https://continuity.datto.com/help/Content/kb/unified-continuity/siris-alto-nas/360040893811.html

Although with Kaseya in charge, I wouldn't be surprised they moved all development in-house and abandoned this github page.

**Comment by ScottMonolith** [comment]
> According to Datto article it's supposed to be supported. https://continuity.datto.com/help/Content/kb/unified-continuity/siris-alto-nas/360040893811.html
> 
> Although with Kaseya in charge, I wouldn't be surprised they moved all development in-house and abandoned this github page, hence the issue remains open.

Yes, Ubuntu 22.04 should be supported based on that article.  The agent does run and does backup successfully but causes a lot of memory consumption.

After keeping the dlad service stopped I can confirm the huge amount of memory usage has also stopped.

I've already had the MSP that manages Datto for us open a ticket.  I found this bug and it's definitely a problem still.

**Comment by Sir-Alex-L** [comment]
So I tried Debian 11 and it crashed on me within 1 day! OOM killed it with over 8GB of RAM usage. I only had 8GB on the machine.

**Comment by Sir-Alex-L** [comment]
Well this was never fixed....

> We are aware an issue where DLA running on Ubuntu 20.04 and Debian 11 causes memory usage to bloat which may result in performance issues on the protected machine
> 
> To mitigate these symptoms you can reboot the protected machine to free the memory usage.
>  
> Currently there is no permanent workaround that our Support team can apply to avoid the original behavior, but we have made our Engineering team aware of this issue and are working together so that a permanent fix can be applied in the future.

**Comment by ScottMonolith** [comment]
> Well this was never fixed....

Indeed, 22.04 and likely Debian 12 as well also suffer from this memory leak bug.

Did the support rep provide a place for you to track progress on the bug?  Or should we just wait for a new version of the Datto Linux agent?  Hopefully it won't be another 3 years for the next release...

**Comment by Sir-Alex-L** [comment]
Told me they'll let me know when the fix is in place, no ETA
Funny thing, Elastio fixed the issue already months ago, they already have support for kernel 6.0 
I wonder why they just won't merge that fork. https://github.com/elastio/elastio-snap

**Comment by ScottMonolith** [comment]
> Told me they'll let me know when the fix is in place, no ETA Funny thing, Elastio fixed the issue already months ago, they already have support for kernel 6.0 I wonder why they just won't merge that fork. https://github.com/elastio/elastio-snap
> 
> I tried compiling that driver with replacing every function call back to dattobd name. It compiles and runs. The issue is dlad service does some driver verification which doesn't pass.

Wow, I had no idea Datto's agent was forked by another company... that repo looks to be _much_ more active than Datto's too.  I doubt they'll be able to just merge all of the changes, but that does give me hope they'll be able to fix it sooner rather than later.

**Comment by Sir-Alex-L** [comment]
> > Told me they'll let me know when the fix is in place, no ETA Funny thing, Elastio fixed the issue already months ago, they already have support for kernel 6.0 I wonder why they just won't merge that fork. https://github.com/elastio/elastio-snap
> > I tried compiling that driver with replacing every function call back to dattobd name. It compiles and runs. The issue is dlad service does some driver verification which doesn't pass.
> 
> Wow, I had no idea Datto's agent was forked by another company... that repo looks to be _much_ more active than Datto's too. I doubt they'll be able to just merge all of the changes, but that does give me hope they'll be able to fix it sooner rather than later.

Eh, not sure, that fork has been active for 2 years now. For some reason Datto refused to clone or merge and relied on recreating the wheel. Either they were unaware of the fork, even though it was mentioned a few times in the past #278 
and #265. Or for some philosophical or legal reason they haven't touched it. Even though both are GPL 2.0 License

**Comment by Swistusmen** [comment]
@Sir-Alex-L We have implememnted some improvement in this topic together with commit 2ad52cb2f0cf8c347083ca84254e89908385ff9e. You can check it out and share how does it works for you

**Comment by Swistusmen** [comment]
@madgamer98 thanks for verification. Are you able to share steps which let you verify this behavior so we could reproduce it? We were just running backups with Datto Linux Agent and was looking how much memory has been taken, never gone to OOM

**Comment by Swistusmen** [comment]
It's great to hear everything works for you. Regarding this I am closing this issue, thanks for your cooperation

**Comment by ScottMonolith** [comment]
@Swistusmen - do you know when a new version of the agent will be released?

Seems I need at least 0.11.3, but on the Datto website the agent version is very different - I show 3.0.16.0 as the agent version.

I need an official release to install on my production systems, but I would love to get Datto back up and running for my Linux hosts!

Thanks!

**Comment by Swistusmen** [comment]
@ScottMonolith Hi, I would prefer not to talk about official agent release dates because I may be not allowed to, I'm sorry. May I ask why do you need new DLA release? Is it about some OS which is unsupported yet, or you are waiting for some improvement?

**Comment by ScottMonolith** [comment]
@Swistusmen that is fair that you cannot talk about official agent release dates.

But my problem is this memory leak!  If I leave Datto running for more than a week the my 20.04 and 22.04 servers start swapping and physical memory is completely consumed.  The only solution I've found is a reboot which is not ideal.

This has a massive negative impact on our production servers, and hence I have disabled Datto on these servers until this bug is resolved.  Sounds like the bug is resolved, but I need an official agent release to be able to deploy it to my production servers.
